text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,595 @@
+to notify the application that no message is available for processing), one
+message, or multiple messages (a batch of messages). "Deliver" and "Receive"
+spans should link to the "Create" or "Publish" span of the messages forwarded,
+thus those spans can link to zero, one, or multiple "Create" spans.
+
+If the operation covered by "Deliver" or "Receive" forwards exactly one
+message, in addition to creating a link to the "Create" span, the "Create" span
+may also be used as a parent of the "Deliver" or "Receive" span. While this is
+not possible for all scenarios (e. g. when receiving a batch of messages, or
+when parenting "Deliver" or "Receive" spans to an ambient context), it can
+improve the user experience in some scenarios.
+
+#### Settlement
+
+Messages can be settled in a variety of different ways. In some cases, messages
+are not settled at all (fire-and-forget), in other cases settlement operations
+are triggered manually by the user, and in callback scenarios settlement can be
+automatically triggered by messaging SDKs based on return values of callbacks.
+
+A "Settle" span should be created for every settlement operation, no matter if
+the settlement operation was manually triggered by the user or automatically
+triggered by SDKs. SDKs will in some cases auto-settle messages in
+push-scenarios when messages are delivered via callbacks. In cases where it is
+possible, it is recommended to create the "Settle" span in the scope of the
+"Deliver" span.
+
+Alternatively, an event can be created instead of a "Settle" span. Events could
+be added to "Deliver" spans or to ambient spans.
+
+"Settle" spans may link to "Create" or "Publish" spans of the messages that are
+settled, however, for some settlement scenarios this is not feasible or
+possible.
+
+## Proposed changes and additions to the messaging semantic conventions
+
+### Operation name
+
+The following operations related to messages are covered by these semantic
+conventions:
+
+| Operation name | Description |
+|----------------|-------------|
+| `publish`      | One or more messages are provided for publishing to an intermediary. |
+| `create`       | A message is created. |
+| `receive`      | One or more messages are requested by a consumer. |
+| `deliver`      | One or more messages are passed to a consumer. |
+| `settle`       | One or more messages are settled. |
+
+For further details about each of those operations refer to the [section about trace structure](#trace-structure).
+
+### Span kind
+
+[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
+SHOULD be set according to the following table, based on the operation a span describes.
+
+| Operation name | Span kind|
+|----------------|-------------|
+| `publish`      | `PRODUCER`, if no `create` spans are present. `INTERNAL` otherwise. |
+| `create`       | `PRODUCER` |
+| `receive`      | `CONSUMER` |
```

Comment:
we should adjust span kind definitions on the spec: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind

currently producer/consumers imply parent-child relationships, which are optional in terms of this semconv

Reply:
There's an issue submitted for this: https://github.com/open-telemetry/opentelemetry-specification/issues/3172.

Reply:
Add this as part of an "Open issues" section to the OTEP.